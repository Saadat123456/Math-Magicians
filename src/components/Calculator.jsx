import React, { useState } from 'react';
import calculate from '../logic/calculate';
import CalculatorButtons from './CalculatorButtons';
import Result from './result';

function Calculator() {
  const [calculations, setCalculations] = useState({
    total: 0,
    next: null,
    operation: null,
  });

  const [result, setResult] = useState('0');

  const createResultString = (calculations) => {
    let result = '';
    if (calculations.total) {
      result += `${calculations.total}`;

      if (calculations.operation) {
        result += ` ${calculations.operation}`;

        if (calculations.next) {
          result += ` ${calculations.next}`;
        }
      }
    } else if (calculations.next) {
      result += ` ${calculations.next}`;
    } else {
      result = '0';
    }
    return result;
  };

  const changeResult = async (buttonName) => {
    setCalculations((prevCalculations) => {
      if (buttonName === '=') {
        if (!prevCalculations.next || !prevCalculations.total) {
          return prevCalculations;
        }
      }
      const newCalculation = calculate(calculations, buttonName);
      setResult(createResultString(newCalculation));
      return newCalculation;
    });
  };

  return (
    <div>
      <Result result={result} className="result" />
      <CalculatorButtons onClick={changeResult} />
    </div>
  );
}

export default Calculator;
