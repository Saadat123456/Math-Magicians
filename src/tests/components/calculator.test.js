import Calculator from '../../components/Calculator';
import React from 'react';
import renderer from 'react-test-renderer';

it('renders correctly', () => {
    const tree = renderer
      .create(<Calculator ></Calculator>)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });