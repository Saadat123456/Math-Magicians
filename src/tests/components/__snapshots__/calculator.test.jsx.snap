// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="calculator-layout"
>
  <div
    className="result"
  >
    0
  </div>
  <ul
    className="row"
  >
    <li
      className=""
    >
      <button
        onClick={[Function]}
        type="button"
      >
        AC
      </button>
    </li>
    <li
      className=""
    >
      <button
        onClick={[Function]}
        type="button"
      >
        +/-
      </button>
    </li>
    <li
      className=""
    >
      <button
        onClick={[Function]}
        type="button"
      >
        %
      </button>
    </li>
    <li
      className=""
    >
      <button
        onClick={[Function]}
        type="button"
      >
        รท
      </button>
    </li>
  </ul>
  <ul
    className="row"
  >
    <li
      className=""
    >
      <button
        onClick={[Function]}
        type="button"
      >
        7
      </button>
    </li>
    <li
      className=""
    >
      <button
        onClick={[Function]}
        type="button"
      >
        8
      </button>
    </li>
    <li
      className=""
    >
      <button
        onClick={[Function]}
        type="button"
      >
        9
      </button>
    </li>
    <li
      className=""
    >
      <button
        onClick={[Function]}
        type="button"
      >
        x
      </button>
    </li>
  </ul>
  <ul
    className="row"
  >
    <li
      className=""
    >
      <button
        onClick={[Function]}
        type="button"
      >
        4
      </button>
    </li>
    <li
      className=""
    >
      <button
        onClick={[Function]}
        type="button"
      >
        5
      </button>
    </li>
    <li
      className=""
    >
      <button
        onClick={[Function]}
        type="button"
      >
        6
      </button>
    </li>
    <li
      className=""
    >
      <button
        onClick={[Function]}
        type="button"
      >
        -
      </button>
    </li>
  </ul>
  <ul
    className="row"
  >
    <li
      className=""
    >
      <button
        onClick={[Function]}
        type="button"
      >
        1
      </button>
    </li>
    <li
      className=""
    >
      <button
        onClick={[Function]}
        type="button"
      >
        2
      </button>
    </li>
    <li
      className=""
    >
      <button
        onClick={[Function]}
        type="button"
      >
        3
      </button>
    </li>
    <li
      className=""
    >
      <button
        onClick={[Function]}
        type="button"
      >
        +
      </button>
    </li>
  </ul>
  <ul
    className="row"
  >
    <li
      className="zero"
    >
      <button
        onClick={[Function]}
        type="button"
      >
        0
      </button>
    </li>
    <li
      className=""
    >
      <button
        onClick={[Function]}
        type="button"
      >
        .
      </button>
    </li>
    <li
      className="equalBtn"
    >
      <button
        onClick={[Function]}
        type="button"
      >
        =
      </button>
    </li>
  </ul>
</div>
`;
